docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,403 @@
+// pitfalls.
+int ecall_with_user_check3(void* ptr) {
+
+    blob_t* blob;
++   void* volatile captured_data;
++   volatile size_t captured_size;
+
+    // Ensure passed-in pointer is not null and the memory block is located
+    // entirely outside of the enclave.
+    if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
+        return -1;
+
+    blob = (blob_t*)ptr;
+
++   // Capture blob descriptors to avoid TOCTOU problems.
++   captured_data = blob->data;
++   captured_size = blob->size;
+
+    // Blobs to be rendered are limited to 4K.
+-   if (blob->size > 4096) //TOCTOU: First fetch
++   if (captured_size > 4096)
+        return -1;
+
+-   return render_data(blob->data, blob->size); //TOCTOU: Second fetch
++   return render_data(captured_data, captured_size);
+}
+```
+
+## 1.4 Ensure proper bounding of data buffers
+
+Another element of marshaling code is ensuring, as complex structures are
+parsed, that nested structures are also within the outer bounds. This is
+especially important for internal functions like `render_data()` that may be
+unaware that the memory is untrusted. Use the same techniques on the inner
+structures that were used on the outer.
+
+> A warning about legacy data-parsers: Take care when passing "hot" data to code
+that may not have been written to parse maliciously crafted input. We have seen
+cases where legacy code that parses self-formatting or self-referencing data was
+used in new enclave applications. This can lead to significant vulnerabilities
+if the data being parsed is still controlled by the host, as may be the case
+with custom marshaling. In the case of this sample code, as mentioned earlier,
+`render_data()` parses "flat" data that is not nested nor self-referencing, so
+it's safe (and performant) to pass it "hot" data that is controlled by the host.
+Care must be taken though, to avoid data-consistency and other app-level
+integrity problems.
+
+```diff
+int ecall_with_user_check4(void* ptr) {
+
+    blob_t* blob;
+    void* volatile captured_data;
+    volatile size_t captured_size;
+
+    // Ensure passed-in pointer is not null and the memory block is located
+    // entirely outside of the enclave.
+    if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Capture blob descriptors to avoid TOCTOU problems.
+    captured_data = blob->data;
+    captured_size = blob->size;
+
+    // Blobs to be rendered are limited to 4K.
+    if (captured_size > 4096)
+        return -1;
+
++   // Ensure validity of nested structure
++   if (!oe_is_outside_enclave(captured_data, captured_size))
++       return -1;
+
++   // Data parsed by render_data is "flat", non-self-referencing.
+    return render_data(captured_data, captured_size);
+}
+```
+## 1.5 Corrected enclave ecall with custom marshaling
+
+Thanks for sticking with us. Here is our corrected ecall function that marshals
+the parameter correctly for SGX-based enclaves:
```

Comment:
```suggestion
Thanks for sticking with us. Here is our updated ecall function that marshals
the parameter correctly for SGX-based enclaves (but is still incorrect for other TEEs
that do not allow enclave code to directly read host memory):
```

Reply:
When I generalize the memory bounds discussion, I'm not going mention "host", which, as you have pointed out, is SGX-centric. Instead, memory buffers will be referred generically as either "trusted" or "untrusted" from the point of view of the enclave. Whether memory is strictly "within" or "outside" the enclave doesn't apply to all hardware implementations. (It is unfortunate that these function names don't map well to this generalized Open Enclave memory model.) Data in buffers that are "untrusted" -- or "outside" -- the enclave must be handled with care. The untrusted buffer contents may be maliciously constructed and may change from read to read.